### DISCUSSION TITLE
advice for using flux2

### DISCUSSION THREAD

**Discussion by jack-evans** [fluxcd/flux2#1530] [Q&A]
Hello,

I am looking for some guidance on how I migrate from our flux v1 stable architecture to flux 2.. Currently, our pipeline and clusters are very much controlled through a central monorepo that holds our helm charts, various releases for those helm charts, and clusters. Our app CI pipeline on master build creates a new docker image and using flux automation we push that onto the dev/master branch of our config repo

Directory structure looks like the following:

```
config
  /charts
     component-a/
        - ....
     component - b/ 
        - ....
  /releases
    - component-a.yaml
    - component-b.yaml
  /clusters
    prod/
      prod-1 /
        - values.yaml
      prod-2 /
        - values.yaml
     test/
      test-1 /
        - values.yaml
      test-2 /
        - values.yaml
      test-3 /
        - values.yaml
```

Currently, the way we move our code from one stage in the pipeline to the next is to shift (via automation) the commits on one branch to the next e.g. dev(master) -> test -> staging -> prod... What I'm slightly concerned by is the use of flux bootstrap, it seems to me to be a great tool and actually removes a lot of the custom tooling my team wrote to essentially do the same thing but for flux v1. 

The current command I am working with is: 

```
flux bootstrap github --owner xxxx --repository config --interval 5m --branch test --components source-controller,kustomize-controller,helm-controller --watch-all-namespaces --hostname xxxx --ssh-hostname xxxx --path ./clusters/test/test-1
```

`flux bootstrap` also creates commits on the branch you provide when you either run the command for the first time or when you update to a newer version of flux.

This presents a problem as at every promotion (e.g. from staging -> prod) we would want to run the above command to be consistent across envs and automate the maintenance of flux 2 (for example, updating from v1.0.0 to v1.0.1). In either of the scenarios mentioned, we would then have to cherry pick those commits back from our prod branch all the way through to dev/master which is not ideal at all.

My only thoughts to get round this potential issue are:

1. use `flux install --export` to a yaml file and use our own tooling to apply it in the same way we do today with flux 1

2. when we `flux bootstrap` we then merge the changes back to the dev/master branch (even if from the prod branch) - this gets more complicated as there may be new changes on the dev branch and we cannot rebase the prod branch

3. when we `flux bootstrap`, all clusters (regardless of place in the pipeline) use the dev/master branch, we then have a `GitRepository` source that points at the actual branch for the cluster in the pipeline and that references the code/charts to use. Only concern with that would be the potential for our CI pipeline to break as on master builds of our applications, we update the `image: ` in the app release to the new image.

4. when we `flux bootstrap`, all clusters (similar to the above) use a completely separate branch called `infra`, which then gets merged in when there are updates to the flux-system directories for the clusters, we then have a `GitRepository` source that points at the actual branch for the cluster in the pipeline and that is used in the references for the code/charts to use.

Thoughts on the above? To me it seems like 3 or 4 would be the best option

**Comment by jack-evans** [comment]
I don't really see an example of flux2 in a git branch pipeline method... only a single branch, where you then apply overrides at each cluster level

**Comment by kingdonb** [comment]
Hello, there is a lot going on in the question as you have posed it, I'm not sure I can give very direct answers as I don't know enough about your environment or reasons you have chosen to arrange it in a particular way. Professional services and some (occasionally offered as free, community-driven) migration workshops are available if you need more specific help with your migration effort.

Here is a pretty new document which you may not have seen yet that has some guidance about repository structure for Flux:

https://fluxcd.io/docs/guides/repository-structure/

The getting started guide has some references to more specific guidance about how to build structure for multi-app, multi-tenant, or multi-cluster installations that are linked at the end:

https://fluxcd.io/docs/get-started/#multi-cluster-setup

We recommend reading and following both examples as they have some surprising conclusions when the lessons are taken together.

Perhaps you've already found all of this, I wanted to highlight them in case you have not. There are some notable ways I think which your repo structure as you painted it differs from what is recommended in these guides.

In the multi-tenancy guide, a branch is used when different access control is needed because branch protection rules can easily be set differently for different orgs and teams. For example, in `flux2-multi-tenancy` a branch is created `dev-team`, where the developers can write directly to, or merge their PRs to without oversight from the platform team responsible for securing Flux. The main branch is where all infrastructure and `flux-system` deploys come from, but a separate tenant Kustomization and GitRepo reconcile this `dev-team` branch in order to add the tenant workloads from the dev team into the cluster more safely and securely.

In this way, the developer team can push changes directly to the prod cluster if their permissions allow, (or be restricted from making such a change if it is not meant to be permitted by policy). They can safely merge pull requests to the `dev-team` branch without invoking the help or review of the infrastructure team, as governed by policies which your organization can decide for itself to be more or less permissive and trusting. We don't manage the installation of Flux in separate environments in this way, Flux installs are not tenants, but should be considered as infrastructure and are usually meant to be managed by a single team (eg. platform support).

In the dev-team context any admin policies can be enforced on the dev-team with Kyverno rules. This is roughly the justification for hosting the tenant workloads on a separate branch; because the dev teams should be given basically unfettered access to deploy their dev workloads in the restricted tenant environment. This is better than complex schemes like CODEOWNERS which can be harder to manage with policy. Please note, the `dev-team` branch is actually a completely separate branch which does not intersect or ever merge with the `main` branch. It describes resources that are owned by the dev team, (this can include production services.)

If the same software is deployed in dev, staging, and prod, but with a normal expectation that they will often have different release targets and different upgrade lifecycles, from a helm chart for example, then all three environments can refer to the same common definition of a HelmRelease, with `kustomization` base references like `../../apps/my-app`.

Find that covered in the `flux2-kustomize-helm-example` tutorial. You can set differently configured parameters (like an ingress rule to use the staging domain name for staging release), or things like (perhaps if you have staging and prod workloads together on one cluster) a target namespace which should use a staging namespace for the staging environment, or version update automation. See: https://github.com/fluxcd/flux2-kustomize-helm-example/blob/main/apps/staging/podinfo-values.yaml

You may have different configurations for each environment targeting different release version streams, like "only install alpha or beta candidate versions in test" and "automate rc upgrades to staging" or "automate only full semver upgrades to prod" are implemented as a patch for the environment, like in the `flux2-kustomize-helm-example`. Hope I addressed some of your questions. 😄 👐 🤝
